node/db/src/rolling/gc.rs

```diff
@@ -0,0 +1,230 @@
+            }
+
+            if let (Ok(total_size), Ok(current_size), last_reachable_bytes) = (
+                self.db.total_size_in_bytes(),
+                self.db.current_size_in_bytes(),
+                self.last_reachable_bytes.load(atomic::Ordering::Relaxed),
+            ) {
+                let should_collect = if last_reachable_bytes > 0 {
+                    total_size > 2 * last_reachable_bytes
+                } else {
+                    total_size > 0 && current_size * 3 > total_size
+                };
+
+                if should_collect {
+                    if let Err(err) = self.collect_once(tipset).await {
+                        warn!("Garbage collection failed: {err}");
+                    }
+                }
+            }
+        }
+    }
+
+    /// This loop listens on events emitted by `forest-cli db gc` and triggers
+    /// `collect_once`
+    pub async fn collect_loop_event(self: &Arc<Self>) -> anyhow::Result<()> {
+        while let Ok(responder) = self.gc_rx.recv_async().await {
+            let this = self.clone();
+            let tipset = (self.get_tipset)();
+            tokio::spawn(async move {
+                let result = this.collect_once(tipset).await;
+                if let Err(e) = responder.send(result) {
+                    warn!("{e}");
+                }
+            });
+        }
+
+        Ok(())
+    }
+
+    /// ## GC workflow
+    /// 1. Walk back from the current heaviest tipset to the genesis block,
+    /// collect all the blocks that are reachable from the snapshot
+    /// 2. writes blocks that are absent from the `current` database to it
+    /// 3. delete `old` database(s)
+    /// 4. sets `current` database to a newly created one
+    async fn collect_once(&self, tipset: Tipset) -> anyhow::Result<()> {
+        let guard = self.lock.try_lock();
+        if guard.is_err() {
+            anyhow::bail!("Another garbage collection task is in progress.");
+        }
+
+        let start = Utc::now();
+        let reachable_bytes = Arc::new(AtomicUsize::new(0));
+
+        info!("Garbage collection started at epoch {}", tipset.epoch());
+        let db = &self.db;
+        // 128MB
+        const BUFFER_CAPCITY_BYTES: usize = 128 * 1024 * 1024;
+        let (tx, rx) = flume::bounded(100);
+        #[allow(clippy::redundant_async_block)]
+        let write_task = tokio::spawn({
+            let db = db.current();
+            async move { db.buffered_write(rx, BUFFER_CAPCITY_BYTES).await }
+        });
+        walk_snapshot(&tipset, DEFAULT_RECENT_STATE_ROOTS, |cid| {
+            let db = db.clone();
+            let tx = tx.clone();
+            let reachable_bytes = reachable_bytes.clone();
+            async move {
+                let block = db
+                    .get(&cid)?
+                    .ok_or_else(|| anyhow::anyhow!("Cid {cid} not found in blockstore"))?;
+
+                let pair = (cid, block.clone());
+                // Key size is 32 bytes in paritydb
```

Comment:
So, does this work at all with RocksDb? If not, how does Forest behave when RocksDb is used?

Reply:
All functions work with rocksdb, however, I did not test thru the crash resilience story for rocksdb so it might not be able to recover when forest is shutdown improperly. I believe we had crash resilience issue with rocksdb before so I think the potential regression is not introduced by this change

Reply:
So you tested that the GC works with rocksdb? How long does it take? I'm asking because the key size value is hardcoded here.